data/data_utils.py

```diff
@@ -0,0 +1,599 @@
+        'boil': '1-1',
+        'melt': '1-2',
+        'freeze': '1-3',
+        'change-the-state-of-matter-of': '1-4',
+        'use-thermometer': '2-1',
+        'measure-melting-point-known-substance': '2-2',
+        'measure-melting-point-unknown-substance': '2-3',
+        'power-component': '3-1',
+        'power-component-renewable-vs-nonrenewable-energy': '3-2',
+        'test-conductivity': '3-3',
+        'test-conductivity-of-unknown-substances': '3-4',
+        'find-living-thing': '4-1',
+        'find-non-living-thing': '4-2',
+        'find-plant': '4-3',
+        'find-animal': '4-4',
+        'grow-plant': '5-1',
+        'grow-fruit': '5-2',
+        'chemistry-mix': '6-1',
+        'chemistry-mix-paint-secondary-color': '6-2',
+        'chemistry-mix-paint-tertiary-color': '6-3',
+        'lifespan-longest-lived': '7-1',
+        'lifespan-shortest-lived': '7-2',
+        'lifespan-longest-lived-then-shortest-lived': '7-3',
+        'identify-life-stages-1': '8-1',
+        'identify-life-stages-2': '8-2',
+        'inclined-plane-determine-angle': '9-1',
+        'inclined-plane-friction-named-surfaces': '9-2',
+        'inclined-plane-friction-unnamed-surfaces': '9-3',
+        'mendellian-genetics-known-plant': '10-1',
+        'mendellian-genetics-unknown-plant': '10-2'
+    }
+    return task_table.get(task_name)
+
+
+def add_current_objects(task_id, look, objects, limit=20):
+    if True:
+        things = re.findall(r'a .*?\n', look.replace(',', '\n').replace(
+            '.', '\n').replace('(', '\n').replace(')', ' ').replace(' an ', ' a '), re.I)
+    else:
+        things = re.findall(r'a .*?\n', look.replace(',',
+                            '\n').replace('.', '\n').replace(' an ', ' a '), re.I)
+
+    flag = 0
+    for thing in things:
+        if 'door' in thing:
+            continue
+
+        if ('(' in thing) and (')' not in thing):
+            start = things.index(thing)
+            flag = 1
+
+        if (')' in thing) and ('(' not in thing):
+            end = things.index(thing)
+            thing = 'and '.join(things[start: end + 1])
+            flag = 0
+
+        if flag:
+            continue
+
+        if clean(thing).strip().strip('.') not in objects:
+            objects.append(clean(thing).strip().strip('.'))
+
+    while len(objects) > limit:  
+        objects.pop(0)
+    return
+
+
+def sanitizeStr(inStr):
+    if inStr is None:
+        return inStr
+    out = inStr.replace("\n\t", " | ").replace("\n", " | ").replace("\t", " | ") 
+    out = out.replace("green house", "greenhouse")
+    out = out.replace("activate bee hive", "open bee hive")
+    out = out.replace("seed peach tree", "peach seed")
+    out = out.replace("orange tree seed", "orange seed")
+    return out
+
+
+def clean(s):
+    clean_toks = ['\n', '\t']
+    for tok in clean_toks:
+        s = s.replace(tok, ' ')
+    return s
+
+
+def downsampling(task_idx_real, curr_task_seq):
+    # Downsampling Task 26 and 29
```

Comment:
This part of the code could have been written by the others but docstrings and comments should have been added for all nontrivial functions. E.g. Why Downsampling has been done only for task 26 and 29?

Reply:
because only these two have a lot of variations. In scienceworld the number of variations of the tasks are different. So it is ok.